src/server/debug/server/server.go

```diff
@@ -101,9 +102,74 @@ func NewDebugServer(env Env) debug.DebugServer {
 		logLevel:      log.LogLevel,
 		grpcLevel:     log.GRPCLevel,
 	}
+	ticker := time.NewTicker(3 * time.Second)
+	query := `SELECT current_timestamp - query_start as runtime,
+                         datname,
+                         usename,
+                         client_addr,
+                         backend_xid,
+                         query
+		  FROM pg_stat_activity WHERE state = 'idle' AND current_timestamp - query_start > interval '10 seconds' ORDER BY runtime DESC;`
+	ctx := pctx.Background("debug server")
```

Comment:
If you allocate a background context here, without a cancel func, there’s no way to cleanly cancel the background logging goroutine (say, completing the write of a log line).

You need two pieces for this to work: a way for another goroutine to signal the background goroutine to terminate; and a way for that goroutine to indicate that it successfully terminated.

The first piece is pretty easy: rather than allocating a background context here, pass in a context to `NewDebugServer` (it looks like each call site either receives a context, or _could_ receive a context, so this is a quick change).

The second piece is a bit more work, and there are a couple of approaches. One would be for the debug server to _not_ run a background goroutine, but for its caller to at some point call something like `debugServer.Run(ctx)` (which would perform the ticker loop). Another approach is to spawn the background goroutine, but return e.g. and error group to the caller, who can then wait for it to terminate after cancelling the context. One can also just return a done channel.

If you adopt this approach, then a clean, orderly server shutdown is possible.